### ISSUE TITLE
Bug: [unified-signatures] Rule does not report some cases due to lack of type information

### ISSUE THREAD

**Issue by StyleShit** [typescript-eslint/typescript-eslint#12505]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.
- [x] If I used AI assistance to prepare this issue, I have reviewed and followed the [AI Contribution Policy](https://typescript-eslint.io/contributing/ai-policy/).

### Playground Link

https://typescript-eslint.io/play/#ts=6.0.2&fileType=.ts&code=C4TwDgpgBAglC8UDOwBOBLAdgcwNwChRIoAhBZNLPffAMwFdMBjYdAe0yloAoBDALgoYcASkEwCDZqw5c%2BglMOxjSkxi3aceAoVRVwA3vihRUEYPVSdeBAL5A&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJygM0sQBNaySgHMmAQ3yxoKdGADa4bDkTRoAe2iQANIqVYl2SCKYbEAEW5cViFvEIA5MQFt%2BABTHRnifCtQZ8cIi62AC%2BwQC6imEhQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
type A = string;
type B = string;

function f(a: string): A;
function f(a: string): B;
function f(a: string): A {
  return a;
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/unified-signatures": [
      "error",
      {
        "ignoreDifferentlyNamedParameters": true
      }
    ]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

`f` should be reported since `A` === `B`

### Actual Result

`f` is not reported because we're checking only the AST without type information

### Additional Info

There are probably more cases like this. It's just the simplest I could come up with. I guess this whole issue is mostly around the question "Do we want to add type information to this rule?"

Co-authored-by: @JoshuaKGoldberg when reviewing #12143 -> #12493 together

**Comment by bradzacher** [comment]
IMO - this is fine. Using different names is a signal we intentionally use to NOT merge things:
```ts
type A = string;
type B = string;

function f(a: string): A;
function f(b: number): A; // does not report as the parameter name is different
function f(a: string | number): A {
  return a.toString();
}


function f2(a: string): A;
function f2(a: number): A; // does report as the parameter name is the same
function f2(a: string | number): A {
  return a.toString();
}
```
[playground](https://typescript-eslint.io/play/#ts=6.0.2&fileType=.ts&code=C4TwDgpgBAglC8UDOwBOBLAdgcwNwChRIoAhBZNLPffAMwFdMBjYdAe0yloAoBDALgoYcASkEwCDZqw5duAI0GZ6AW3kRUY2JMYt2nHgKFUoAHyjK1GrXADe%2BKFFQRg9VJ14A6YGwDKlHG4RAgBfGjpdGQMAJj5BFGFsGx1pfS5Yo0t1TXEUvVlaDPiA7DMLVWybKHtHZ1d3KC8ff0Sg0KA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJygM0sQBNaySgHMmAQ3yxoKdGADa4bDkTRoAe2iQANIqVYl2SCKYbEAEW5cViFvEIA5MQFt%2BABTHRnifCtQZ8cIi62AC%2BwQC6imEhQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

This was built into the rule by request from users. So IMO if the type's are the same but with different names, it is correct to treat them as different types.

**Comment by StyleShit** [comment]
> This was built into the rule by request from users

Maybe it's worth emphasizing this in the docs then

It's kinda inferable from [`ignoreDifferentlyNamedParameters`](https://typescript-eslint.io/rules/unified-signatures/#ignoredifferentlynamedparameters), but I think we should make it more explicit

Also, does it mean function `c` in #12143 should not be reported (as opposed to what OP is saying)?

**Comment by JoshuaKGoldberg** [comment]
+1 to making this more explicit. It's also surprising to me as a user of this who likes things being strict. Added +1 to:

> IMO if the type's are the same but with different names, it is correct to treat them as different types.

Shouldn't this be, like, the default? With opt-out for folks who don't want to be so strict?

**Comment by bradzacher** [comment]
Having an opt-out would mean augmenting the rule with type information - which would be a breaking change with a decent amount of impact.

If it was cheap to optionally add type info - I'd agree with you - setting a flag to report on this would be the thing to do!
I.e. A sister option to `ignoreDifferentlyNamedParameters` would be good if we could cheaply integrate type info.

Though I'd _personally_ argue that such a feature request would fail the niche-ness test - the rule is almost a decade old at this point and this is the first time someone has noticed and reported this case - from that it follows that its not likely something a large enough portion of the community cares about.

With those two combined I'm _personally_ -1 on such a request - edge case value with the cost of a breaking change doesn't seem worth it, IMO.

**Comment by StyleShit** [comment]
Ok, now Brad made me -0.5 on this 😬

**Comment by JoshuaKGoldberg** [comment]
Seems like an easy close then 😄. I'm not passionate about it myself.
